include/alp/utils/parser/MatrixFileReader.hpp

```diff
@@ -0,0 +1,161 @@
+
+/*
+ *   Copyright 2021 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * @author A. N. Yzelman
+ * @date 25th of May, 2017
+ */
+
+#ifndef _H_MATRIXFILEREADER
+#define _H_MATRIXFILEREADER
+
+#include <algorithm> //std::max
+#include <iostream>
+#include <string>
+#include <type_traits>
+
+#include "MatrixFileIterator.hpp"
+#include "MatrixFileProperties.hpp"
+#include "MatrixFileReaderBase.hpp"
+
+namespace alp {
+	namespace utils {
+
+		template< typename T, typename S = size_t >
+		class MatrixFileReader : public internal::MatrixFileReaderBase< T, S > {
+
+			static_assert( std::is_integral< S >::value, "The template parameter S to MatrixFileReader must be integral." );
+
+			template< typename U, typename V >
+			friend std::ostream & operator<<( std::ostream & out, const MatrixFileReader< U, V > & A );
+
+		private:
+
+
+		public:
```

Comment:
`public` should be indented one tab to the right. Also, the code below it should also be indented one tab more.